map to unordered_map for augmenting sparsity

lindsayad opened the pull request:
`std::map::operator[]` was really punishing us in profiling so I think
it makes sense to switch to a container whose lookup is on average
constant instead of log(N).

On a case given to me by @snschune the change here took me from

![dont-solve](https://user-images.githubusercontent.com/10248304/125847219-41df539f-ee93-4044-9fa4-3504b8f35dc9.png)

to

![half-change-structs](https://user-images.githubusercontent.com/10248304/125847232-446fec93-43fa-4599-bc2b-5ed1ced030d4.png)

which is a 35% reduction in the timing of `NonlinearSystemBase::findImplicitGeoemtricCouplingEntries`

Refs #18353

lindsayad committed:
map to unordered_map for augmenting sparsity

`std::map::operator[]` was really punishing us in profiling so I think
it makes sense to switch to a container whose lookup is on average
constant instead of log(N).

Refs #18353

roystgnr commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -274,9 +274,9 @@ class NonlinearSystemBase : public SystemBase, public PerfGraphInterface
-                                       std::map<dof_id_type, std::vector<dof_id_type>> & graph);
+  void findImplicitGeometricCouplingEntries(
+      GeometricSearchData & geom_search_data,
+      std::unordered_map<dof_id_type, std::vector<dof_id_type>> & graph);
```
Since we think this is only used internally, and since we're about to really mess with anyone crazy enough to be calling it from third-party code, should we make that policy official and move this to `protected:` (or even `private:`??) rather than `public:`, while we're at it?

lindsayad replied:
I like that plan!

lindsayad committed:
Make findImplicitGeometricCouplingEntries private

We'll discover if any of our apps try to call this. I searched Griffin,
BISON, and THM which are kind of our power apps and didn't see anything

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/778402/) on 718bf15 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18354/site)

This comment will be updated on new commits.

moosebuild commented:
Job [External App tests](https://civet.inl.gov/job/778405/) on 718bf15 : invalidated by @lindsayad

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/778406/) on 718bf15 : invalidated by @lindsayad

moosebuild commented:
Job [Moose Install Target Tests](https://civet.inl.gov/job/778414/) on 718bf15 : invalidated by @lindsayad

lindsayad commented:
@roystgnr could I get a re-approval here?
